### PR TITLE
Split infer info data structure in strings

### PR DESCRIPTION
**Pull request by ajreynol**
Also refactors the mkExplain method and adds some const to some arguments.

### CODE REVIEW

File: src/theory/strings/inference_manager.h [hunk 1]
```diff
@@ -125,16 +126,22 @@ class InferenceManager
...
                      bool asLemma = false);
   /** same as above, but where exp_n is empty */
-  void sendInference(std::vector<Node>& exp,
+  void sendInference(const std::vector<Node>& exp,
                      Node eq,
                      const char* c,
                      bool asLemma = false);
+  /** Send inference
+   *
+   * Calls the appropriate call to send inference based on the infer info
```
Comment (4tXJ7f) [inline src/theory/strings/inference_manager.h]: "Calls the appropriate call" reads a bit weird. Maybe just refer to the documentation of `sendInference`?

File: src/theory/strings/theory_strings.cpp [hunk 2]
```diff
@@ -4110,51 +4088,57 @@ Node TheoryStrings::mkLength( Node t ) {
...
+    if (std::find(ap.begin(), ap.begin() + i, api) != ap.begin() + i)
+    {
+      // already processed
+      continue;
+    }
+    if (api.getKind() == AND)
+    {
+      for (const Node& apic : api)
+      {
+        ap.push_back(apic);
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]: I am not a huge fan of modifying `ap` while iterating over it. The for loop above makes it seem like the size is fixed. Why not do two loops? Instead of copying `a` to `ap` in the beginning, iterate over `ap` and flatten the `AND`s into `ap`. This shouldn't be less efficient (in fact you save copying some useless elements into `ap`).

File: src/theory/strings/theory_strings.cpp [hunk 3]
```diff
@@ -4110,51 +4088,57 @@ Node TheoryStrings::mkLength( Node t ) {
...
+    }
+    Debug("strings-explain") << "Add to explanation " << api << std::endl;
+    if (api.getKind() == NOT && api[0].getKind() == EQUAL)
+    {
+      Assert(hasTerm(api[0][0]));
+      Assert(hasTerm(api[0][1]));
+      // ensure that we are ready to explain the disequality
+      AlwaysAssert(d_equalityEngine.areDisequal(api[0][0], api[0][1], true));
     }
+    Assert(api.getKind() != EQUAL || d_equalityEngine.areEqual(api[0], api[1]));
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]: One detail that is different here is that we are not skipping cases where `api[0] == api[1]`. Is this change intentional?

Comment (ajreynol) [reply src/theory/strings/theory_strings.cpp]: Yes, I realized that the core `explain` method is robust to this case, so this was being checked twice.

File: src/theory/strings/theory_strings.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/strings/theory_strings.cpp]: The flattening of AND may need to be done recursively, which makes this a bit of a pain.  I'll make a utility to do this.

File: src/theory/strings/theory_strings_utils.cpp [hunk 4]
```diff
@@ -44,6 +44,21 @@ Node mkAnd(std::vector<Node>& a)
   return NodeManager::currentNM()->mkNode(AND, au);
 }
 
+void getConjuncts(Node n, std::vector<Node>& conj)
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings_utils.cpp]: Sorry for being annoying but how about making this function take a `Kind` to flatten and making it non-recursive?

File: src/theory/strings/theory_strings.cpp (hunk 2 above)
Comment (4tXJ7f) [reply src/theory/strings/theory_strings.cpp]: Yeah, but it is worth it IMHO :) The updated code is much easier to read.

File: src/theory/strings/theory_strings_utils.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/strings/theory_strings_utils.cpp]: Done.  Notice that this is pretty performance critical code, so I've made a special case for when the top symbol is not the kind, in which we don't do a traversal.

File: src/theory/strings/theory_strings.cpp [hunk 5]
```diff
@@ -4110,50 +4088,44 @@ Node TheoryStrings::mkLength( Node t ) {
...
-        for( unsigned j=ps; j<antec_exp.size(); j++ ) {
-          Debug("strings-explain") << "  " << antec_exp[j] << std::endl;
-        }
-        Debug("strings-explain") << std::endl;
-      }
+  // copy to processing vector
+  std::vector<Node> aconj;
+  for (const Node& ac : a)
+  {
+    utils::flattenOp(ac, aconj);
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]:
```suggestion
    utils::flattenOp(AND, ac, aconj);
```

File: src/theory/strings/theory_strings_utils.cpp [hunk 6]
```diff
@@ -44,6 +44,45 @@ Node mkAnd(std::vector<Node>& a)
   return NodeManager::currentNM()->mkNode(AND, au);
 }
 
+void getConjuncts(Kind k, Node n, std::vector<Node>& conj)
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings_utils.cpp]:
```suggestion
void flattenOp(Kind k, Node n, std::vector<Node>& conj)
```
